src/transformations/tinshift_impl.hpp

```diff
@@ -410,6 +425,29 @@ Evaluator::Evaluator(std::unique_ptr<TINShiftFile> &&fileIn)
 
 // ---------------------------------------------------------------------------
 
+static inline double sqr(double x) { return x * x; }
+static inline double squared_distance(double x1, double y1, double x2,
+                                      double y2) {
+    return sqr(x1 - x2) + sqr(y1 - y2);
+}
+static double distance_point_segment(double x, double y, double x1, double y1,
+                                     double x2, double y2) {
+    // squared distance of point x/y to line segment x1/y1 -- x2/y2
+    double t = ((x - x1) * (x2 - x1) + (y - y1) * (y2 - y1)) /
```

Comment:
in case of degenerate triangles, a division by zero could occur

Reply:
> Should the parser maybe throw an error if any of the input triangles in degenerate?

maybe, but we want the initial parsing to be as fast as possible. Anyway, this is not a critical point. That can be left for later potentially

Reply:
Okay, I added code that avoids division by zero and tests for the two cases of degenerate triangles (one side is zero length or two angles are 0°). This also made me fix another bug: if the list of triangles is zero, the function would try to access the non-existant triangle at position 0. Whoops. XD I added a test for that as well.